src/terrain/terrain.hpp

```diff
@@ -353,6 +410,12 @@ class Terrain {
     Terrain(const std::string path, const std::map<int, const Material>* material);
 
     // TODO place block
+
+    inline path::AdjacentIterator
+    get_tile_adjacent_iterator(size_t pos, UnitPath path_type) const {
+        return path::AdjacentIterator(*this, pos, path_type);
```

Comment:
Aren't iterators supposed to be a subclass?
You can do this with `using iterator = path::AdjacentIterator;` in the public category

Also begin/end?

Reply:
I have no idea what you are talking about. You should give some instructions.

Reply:
When a `vector` returns an `iterator`, its of type `vector<T>::iterator`.
A `terrain` should return a `terrain::iterator`.

So add this:

```c++
class Terrain {
 public: 
    // Stuff...
    using iterator = path::AdjacentIterator; 

    inline iterator 
    get_tile_adjacent_iterator(size_t pos, UnitPath path_type) const {
        return iterator(*this, pos, path_type);
    }
}
```

Also, typically iterators are returned by the `begin()`/`end()` methods, like in `vector`. Why doesn't `Terrain` do the same thing?

Reply:
the iterator always exists for vectors, but here each iterator is generated when the user chooses a tile. In this implimentation begin is set implicitly.